beets: 1.4.9 -> unstable-2020-12-22

lovesegfault opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change
During https://github.com/NixOS/nixpkgs/pull/90504 it became clear that `bs1770gain` is a menace to our users, given that it comes bundled with white supremacist messages. The only user of it is beets, which requires it for replaygain functionality in version 1.4.9.

The current HEAD of beets has added support for using ffmpeg for replaygain, and so this updates beets to the HEAD which removes the dependency on `bs1770gain` altogether. We also remove `bs1770gain` from the tree, as it now has no maintainers and no reverse dependencies.

I attempted to copy the stable drv as closely as possible, any cleanups and the like will follow in a later PR.

cc. @flokli, @lassulus, @doronbehar @mweinelt 

###### Things done
<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

lovesegfault committed:
beets: 1.4.9 -> unstable-2020-12-22

The maintainer has stopped cutting releases[1]. Since the last release,
1.4.9, includes a dependency that is filled with hate speech[2] it's all
the more reason to package the unstable version and eliminate that
requirement.

Moreover a number of fixes, improvements, and features have landed
since.

[1]: https://github.com/beetbox/beets/issues/3625
[2]: https://github.com/NixOS/nixpkgs/pull/90504

Co-authored-by: Doron Behar <doron.behar@gmail.com>

flokli commented:
Mhhmm… Any chance we could just cherry-pick https://github.com/beetbox/beets/pull/3056/files?

lovesegfault commented:
> Mhhmm… Any chance we could just cherry-pick https://github.com/beetbox/beets/pull/3056/files?

We could if all we wanted was to get rid of bs1770gain. There is, however, another big issue here which is `beets` basically doesn't do releases anymore, so regardless we should have a `beetsUnstable` package that tracks master. 

It's pretty much the only way of having up-to-date beets at this point.

lovesegfault commented:
Talked to @flokli on IRC. We agreed that it's best to just package beets as unstable until they do another release, instead of packaging stable and unstable, as that's a lot more work.

lovesegfault committed:
beets: add lovesegfault to maintainers

lovesegfault committed:
bs1770gain: nuke

The package is filled with white supremacist hate speech, and the only
reverse dependency on it, beets, can now use ffmpeg instead.

SuperSandro2000 commented:
> Mhhmm… Any chance we could just cherry-pick [beetbox/beets#3056 (files)](https://github.com/beetbox/beets/pull/3056/files)?

FYI https://github.com/NixOS/nixpkgs/pull/90504#issuecomment-690901391

SuperSandro2000 commented on pkgs/top-level/all-packages.nix:
```diff
@@ -20727,8 +20727,6 @@ in
 
   bristol = callPackage ../applications/audio/bristol { };
 
-  bs1770gain = callPackage ../applications/audio/bs1770gain { };
```
I would still prefer if we add a throw alias to not create any eval errors anywhere. The message should just be that the package got removed.

lovesegfault replied:
Agreed, done :)

SuperSandro2000 reviewed (commented):
After you moved the plugins we now have two times plugins in the path. I think it should be dropped from the nix file.

thiagokokada commented:
Result of `nixpkgs-review pr 108275` [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>beets</li>
  </ul>
</details>

lovesegfault commented:
> After you moved the plugins we now have two times plugins in the path. I think it should be dropped from the nix file.

Done :)

SuperSandro2000 commented on pkgs/top-level/aliases.nix:
```diff
@@ -78,6 +78,7 @@ mapAliases ({
   bittorrentSync = throw "bittorrentSync has been deprecated by resilio-sync."; # added 2019-06-03
   bittorrentSync14 = throw "bittorrentSync14 has been deprecated by resilio-sync."; # added 2019-06-03
   bittorrentSync20 = throw "bittorrentSync20 has been deprecated by resilio-sync."; # added 2019-06-03
+  bs1770gain = throw "bs1770gain has been removed from nixpkgs, as it distributed hate speech."; # added 2021-01-02
```
```suggestion
  bs1770gain = throw "bs1770gain has been removed from nixpkgs."; # added 2021-01-02
```
Lets just leave it like this.

SuperSandro2000 commented:
This is a semi-automatic executed nixpkgs-review which does not build all packages (e.g. lumo, tensorflow or pytorch)
If you find some bugs or got suggestions for further things to search or run please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 108275` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
1 package built:
</summary>
<ul>
<li>beets</li>
</ul>
</details>

doronbehar commented on pkgs/tools/audio/beets/default.nix:
```diff
@@ -6,51 +6,57 @@
+, enableBadfiles         ? true, flac ? null, mp3val ? null
+, enableBeatport         ? true
+, enableBpsync           ? true
+, enableConvert          ? true, ffmpeg_3 ? null
```
While at it, let's upgrade to `ffmpeg`.

lovesegfault replied:
Done

doronbehar commented on pkgs/tools/audio/beets/default.nix:
```diff
@@ -109,21 +119,21 @@ let
 in pythonPackages.buildPythonApplication rec {
   pname = "beets";
-  version = "1.4.9";
+  version = "unstable-2020-12-22";
```
```suggestion
  # The stable version 1.4.9 is 1000 behinds master, and includes many bug fixes 
  # and improvements. It also doesn't require bs1770gain. See:
  # - https://github.com/NixOS/nixpkgs/pull/90504
  # - https://discourse.beets.io/t/forming-a-beets-core-team/639
  version = "unstable-2020-12-22";
```

lovesegfault replied:
Done

doronbehar commented on pkgs/tools/audio/beets/default.nix:
```diff
@@ -187,25 +203,16 @@ in pythonPackages.buildPythonApplication rec {
-    # https://github.com/beetbox/beets/pull/3621
-    # Can be dropped after 1.4.9
-    ./compatibility-with-breaking-changes-to-the-ast-module.patch
+    ./replaygain-default-ffmpeg.patch
```
```suggestion
    # Since we don't package bs1770gain, we make the replaygain plugin use ffmpeg by default
    ./replaygain-default-ffmpeg.patch
```

lovesegfault replied:
Done

doronbehar commented on pkgs/tools/audio/beets/default.nix:
```diff
@@ -217,11 +224,7 @@ in pythonPackages.buildPythonApplication rec {
-    ' beetsplug/replaygain.py
-    sed -i -e 's/if has_program.*bs1770gain.*:/if True:/' \
-      test/test_replaygain.py
+    sed -i -e 's,"ffmpeg","${ffmpeg_3.bin}/bin/ffmpeg",' beetsplug/replaygain.py
```
How about removing this part, and adding the patch for replaygain above, with:

```nix
  (substituteAll {
    src = ./ffmpeg-replay-gain.patch;
    ffmpeg = lib.getBin ffmpeg;
  })
```

Ideally perhaps the patch should be applied conditionally if replaygain is enabled.

lovesegfault replied:
Did the `substituteAll`. I dislike conditional patching, so did not do that.

doronbehar commented on pkgs/tools/audio/beets/default.nix:
```diff
@@ -284,7 +284,7 @@ in pythonPackages.buildPythonApplication rec {
     homepage = "http://beets.io";
     license = licenses.mit;
-    maintainers = with maintainers; [ aszlig domenkozar pjones ];
+    maintainers = with maintainers; [ aszlig domenkozar lovesegfault pjones ];
```
Please add me as well :+1: 

```suggestion
   maintainers = with maintainers; [ aszlig domenkozar lovesegfault doronbehar pjones ];
```

lovesegfault replied:
Done.

lovesegfault committed:
beets: add doronbehar to maintainers

doronbehar commented on pkgs/tools/audio/beets/default.nix:
```diff
@@ -215,13 +232,7 @@ in pythonPackages.buildPythonApplication rec {
-    sed -i -e 's,\(util\.command_output(\)\([^)]\+\)),\1[b"${ffmpeg_3.bin}/bin/ffmpeg" if args[0] == b"ffmpeg" else args[0]] + \2[1:]),' beetsplug/convert.py
-  '' + optionalString enableReplaygain ''
-    sed -i -re '
-      s!^( *cmd *= *b?['\'''"])(bs1770gain['\'''"])!\1${bs1770gain}/bin/\2!
```
./replaygain-default-ffmpeg.patch doesn't include the change done here, equivalent for ffmpeg, which was present in a previous version of the PR.

lovesegfault replied:
Okay, I do not want to bother with fixing this now. I will revert to the sed behavior and it can be solved in a subsequent PR.

doronbehar commented on pkgs/tools/audio/beets/default.nix:
```diff
@@ -215,13 +232,7 @@ in pythonPackages.buildPythonApplication rec {
-    ' beetsplug/replaygain.py
-    sed -i -e 's/if has_program.*bs1770gain.*:/if True:/' \
-      test/test_replaygain.py
+    sed -i -e 's,\(util\.command_output(\)\([^)]\+\)),\1[b"${ffmpeg.bin}/bin/ffmpeg" if args[0] == b"ffmpeg" else args[0]] + \2[1:]),' beetsplug/convert.py
```
Ideally this change would have been also part of a patch..

lovesegfault replied:
Same as the above.

doronbehar commented on pkgs/tools/audio/beets/default.nix:
```diff
@@ -215,13 +232,7 @@ in pythonPackages.buildPythonApplication rec {
       s,"mp3val","${mp3val}/bin/mp3val",
```
And the changes here and above as well, part of a patch and not a sed command.. But this should not block the PR if you don't want to do it @lovesegfault .

lovesegfault replied:
Same as the above.

lovesegfault commented:
@doronbehar My goal with this PR is to get `beets` master working and in an acceptable state, not to undertake an overhaul of the drv. While I agree that it's needed, it's for a separate PR.

(Just to clarify why I rejected/marked as resolved the last round of comments)

doronbehar commented:
> @doronbehar My goal with this PR is to get `beets` master working and in an acceptable state, not to undertake an overhaul of the drv. While I agree that it's needed, it's for a separate PR.

Of course, but some of the sed commands are not updated and in fact they do nothing. I pushed a fixup commit, which I'd like you to review :).

lovesegfault commented:
> > @doronbehar My goal with this PR is to get `beets` master working and in an acceptable state, not to undertake an overhaul of the drv. While I agree that it's needed, it's for a separate PR.
> 
> Of course, but some of the sed commands are not updated and in fact they do nothing. I pushed a fixup commit, which I'd like you to review :).

Tests pass, all looks good to me :)

Added you as co-author while rebasing.

doronbehar commented:
Fixed a small comment badly written in the expression.

doronbehar commented:
It'd be nice to get a :+1: from a Darwin user for this, to make sure the darwin build succeeds as well. Also, I don't have currently at hand a functional beets library to test that basic commands work, but I guess we should count upon the tests. Also, for the record, the files in `./result` show up that `substituteAll` patches have worked.

SuperSandro2000 commented:
This is a semi-automatic executed nixpkgs-review which does not build all packages (e.g. lumo, tensorflow or pytorch)
If you find some bugs or got suggestions for further things to search or run please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 108275` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
1 package built:
</summary>
<ul>
<li>beets</li>
</ul>
</details>
